tokenserver/tests/test_service.py

```diff
@@ -415,15 +424,122 @@ def test_client_state_change(self):
         self.assertEqual(res.json['status'], 'invalid-client-state')
         headers['X-Client-State'] = 'aaaa'
         mock_response["idpClaims"]["fxa-generation"] += 1
+        mock_response["idpClaims"]["fxa-keysChangedAt"] += 1
         with self.mock_browserid_verifier(response=mock_response):
             res = self.app.get('/1.0/sync/1.1', headers=headers, status=401)
         self.assertEqual(res.json['status'], 'invalid-client-state')
 
+    def test_fxa_kid_change(self):
+        # Starting off not reporting keys_changed_at.
+        # We don't expect to encounter this in production, but it might
+        # happen to self-hosters who update tokenserver without updating
+        # their FxA stack.
+        headers = {
+            "Authorization": "BrowserID %s" % self._getassertion(),
+            "X-Client-State": "616161",
+        }
+        mock_response = {
+            "status": "okay",
+            "email": "test@mozilla.com",
+            "idpClaims": {"fxa-generation": 1234},
+        }
+        with self.mock_browserid_verifier(response=mock_response):
+            res = self.app.get('/1.0/sync/1.1', headers=headers)
+        token = self.unsafelyParseToken(res.json["id"])
+        self.assertEqual(token["fxa_kid"], "0000000001234-YWFh")
+        # Now pretend we updated FxA and it started sending keys_changed_at.
+        mock_response["idpClaims"]["fxa-generation"] = 2345
+        mock_response["idpClaims"]["fxa-keysChangedAt"] = 2345
+        headers["X-Client-State"] = "626262"
+        with self.mock_browserid_verifier(response=mock_response):
+            res = self.app.get('/1.0/sync/1.1', headers=headers)
+        token = self.unsafelyParseToken(res.json["id"])
+        self.assertEqual(token["fxa_kid"], "0000000002345-YmJi")
+        # If we roll back the FxA stack so it stops reporting keys_changed_at,
+        # users will get locked out because we can't produce `fxa_kid`.
+        del mock_response["idpClaims"]["fxa-keysChangedAt"]
+        with self.mock_browserid_verifier(response=mock_response):
+            self.app.get('/1.0/sync/1.1', headers=headers, status=401)
```

Comment:
Might be worth adding `res.json['status']` checks for these new status=401 tests

Reply:
good idea; updated